text/0097-log-data-model.md

```diff
@@ -0,0 +1,988 @@
+to display the short name in such UI elements.
+
+For example a dropdown list of severities that allows filtering log records by
+severities is likely to be more usable if it contains the short names of
+`SeverityNumber` (and thus has a limited upper bound of elements) compared to a
+dropdown list, which lists all distinct `SeverityText` values that are known to
+the system (which can be a large number of elements, often differing only in
+capitalization or abbreviated, e.g. "Info" vs "Information").
+
+#### Comparing Severity
+
+In the contexts where severity participates less-than / greater-than comparisons
+`SeverityNumber` field should be used. `SeverityNumber` can be compared to
+another `SeverityNumber` or to numbers in the 1..24 range (or to the
+corresponding short names).
+
+When severity is used in equality or inequality comparisons (for example in
+filters in the UIs) the recommendation is to attempt to use both `SeverityText`
+and short name of `SeverityNumber` to perform matches. For example if we have a
+record with `SeverityText` field equal to "Informational" and `SeverityNumber`
+field equal to INFO then it may be preferable from the user experience
+perspective to ensure that **severity="Informational"** and **severity="INFO"**
+conditions both to are TRUE for that record.
+
+### Field: `ShortName`
+
+Type: string.
+
+Description: Short event identifier that does not contain varying parts.
+`ShortName` describes what happened (e.g. "ProcessStarted"). Recommended to be
+no longer than 50 characters. Optional. Not guaranteed to be unique in any way.
+Typically used for filtering and grouping purposes in backends.
+
+### Field: `Body`
+
+Type: any.
+
+Description: A value containing the body of the log record (see the description
+of `any` type above). Can be for example a human-readable string message
+(including multi-line) describing the event in a free form or it can be a
+structured data composed of arrays and maps of other values. Can vary for each
+occurrence of the event coming from the same source.
+
+### Field: `Resource`
+
+Type: key/value pair list.
+
+Description: Describes the source of the log, aka
+[resource](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/overview.md#resources).
+"value" of each pair is of `any` type. Multiple occurrences of events coming
+from the same event source can happen across time and they all have the same
+value of `Resource`. Can contain for example information about the application
+that emits the record or about the infrastructure where the application runs.
+Data formats that represent this data model may be designed in a manner that
+allows the `Resource` field to be recorded only once per batch of log records
+that come from the same source. SHOULD follow OpenTelemetry
+[semantic conventions for Resources](https://github.com/open-telemetry/opentelemetry-specification/tree/master/specification/resource/semantic_conventions).
+
```

Comment:
[Optional] One of the challenges we've been having with logging at Google is log duplication. Log duplication can be triggered in different phases of the pipeline at both client side and server, and it can been tricky to diagnose what the root cause is. For instance, the logging agent has some buffering and retrying mechanism that can potentially contribute to log duplication. One feature that helps with this is to enforce an `insertId` (or UUID, whatever we want to call it).

Quoting the public documentation (https://cloud.google.com/logging/docs/reference/v2/rest/v2/LogEntry):

```
insertId

String. Optional. A unique identifier for the log entry. If you provide a value, then Logging considers other log entries in the same project, with the same timestamp, and with the same insertId to be duplicates which are removed in a single query result. However, there are no guarantees of de-duplication in the export of logs.

If the insertId is omitted when writing a log entry, the Logging API assigns its own unique identifier in this field.

In queries, the insertId is also used to order log entries that have the same logName and timestamp values.
```

Technically speaking, the earlier the `insertID` gets set (ideally by the log producer, but in reality it's quite challenging), the higher chances that we can deduplicate logs or diagnose root cause when log duplication happens.

While we can put `insertID` in the "attribute" section of the proposed data model, this may be a chance to promote the `insertID` concept to the broader audience by making it a top-level field in case others have this challenge and are looking for a similar solution as well.

Reply:
This is something I had contemplated adding but eventually didn't. I am open to adding it but would like to see more opinions and thoughts on this.

Some log sources natively include this (e.g. Windows Event Logs IIRC), many others don't (e.g. when collecting from a file log there is no natural unique id - file name+position is not necessarily unique due to log rotations). Logging libraries also do not typically generate ids for log record.

It is likely that in the majority of cases the id is not going to be available. In this case according to the design guidelines instead of making it a top-level field we should instead define a semantic convention for it and recommend to include in the attributes.

I would like hear more on this topic.

Reply:
Yeah, the reality today is that it's not present in many cases and almost impossible to enforce or even promote from any individual log producer or collector. But the pain is real. I'm also curious how others deal with similar problems.

Reply:
Adding it as an optional field is a nice way to potentially get it used more and more consistently.

Reply:
I agree that having a record if is useful, but I have a hard time defining it correctly.

The uniqueness requirements are not entirely clear for such field. Presumably we want it to be globally unique however it may not be the case for other log formats so we cannot make that a requirement. It is tricky to define the scope within which a record id MUST be unique to make it useful.

If we require that it is unique per Resource it may be enough for deduplication (since supposedly the Resource itself will contain enough information to be globally unique). However it makes deduplication expensive since it requires a lookup using a composite key with multiple components.

Reply:
oklog  blog had an interesting discussion about that https://peter.bourgon.org/ok-log/#ingestion

Reply:
Thanks for the link. ULID are indeed very useful as log record ids since the fulfill the uniqueness (for deduplication) and sortability requirements (for restoring order in out-of-order delivery scenarios - order matters for logs because log record lack explicit linkage and parent/child association concepts that spans have).

However, this only works if we require the that log record IDs are ULID. If we add such requirement we will essentially disallow sources that already have a notion of record id to use it. It means when the log record is translated from another log format to this data model it must be assigned a new ULID and we will ignore the record ID in the original format or will preserve it in an attribute.

Alternatively we can recommend using ULIDs, but not require it. This way readers of log record can check if the record id looks like an ULID they can use for deduplication directly, otherwise they can resort to slower methods that assumes the id is only unique in combination with the Resource.

I'd leave this out from this OTEP for now to give us time to think through. We can add an optional ID field to the Data Model in a follow up OTEP.

Reply:
Recorded an issue to work on this later https://github.com/open-telemetry/opentelemetry-specification/issues/597